src/Client.ts

```diff
@@ -0,0 +1,168 @@
+   * @param supabaseUrl The unique Supabase URL which is supplied when you create a new project in your project dashboard.
+   * @param supabaseKey The unique Supabase Key which is supplied when you create a new project in your project dashboard.
+   * @param options.schema You can switch in between schemas. The schema needs to be on the list of exposed schemas inside Supabase.
+   * @param options.autoRefreshToken Set to "true" if you want to automatically refresh the token before expiring.
+   * @param options.persistSession Set to "true" if you want to automatically save the user session into local storage.
+   * @param options.detectSessionInUrl Set to "true" if you want to automatically detects OAuth grants in the URL and signs in the user.
+   * @param options.headers Any additional headers to send with each network request.
+   */
+  constructor(
+    public supabaseUrl: string,
+    public supabaseKey: string,
+    options?: SupabaseClientOptions
+  ) {
+    if (!supabaseUrl) throw new Error('supabaseUrl is required.')
+    if (!supabaseKey) throw new Error('supabaseKey is required.')
+
+    const settings = { ...DEFAULT_OPTIONS, ...options }
+    this.restUrl = `${supabaseUrl}/rest/v1`
+    this.realtimeUrl = `${supabaseUrl}/realtime/v1`.replace('http', 'ws')
+    this.authUrl = `${supabaseUrl}/auth/v1`
+    this.schema = settings.schema
+
+    this.auth = this._initGoTrueClient(settings)
+    this.realtime = this._initRealtimeClient()
+
+    this.realtime.onOpen(() => console.log('OPEN'))
+    this.realtime.onClose(() => console.log('CLOSED'))
+    this.realtime.onError((e: Error) => console.log('Socket error', e))
+  }
+
+  rpc(tableName: string, params: object): any {
+    return this._initPostgRESTClient().rpc(tableName, params)
+  }
+
+  /**
+   * Perform a table operation.
+   * 
+   * @param table The table name to operate on.
+   */
+  from(tableName: string): any {
+    // At this point, we don't know whether the user is going to
+    // make a call to Realtime or to PostgREST, so we need to do
+    // an intermdiary step where we return both.
+    // We have to make sure "this" is bound correctly on each part.
+    let rest = this._initPostgRESTClient()
+    let subscription = new RealtimeWrapper(this.realtime, this.schema, tableName)
+
+    const builder = {
+      rest,
+      subscription,
+      select: (columns: string | undefined) => {
+        return rest.from(tableName).select(columns)
+      },
+      insert: (values: any, options?: any) => {
+        return rest.from(tableName).insert(values, options)
+      },
+      update: (values: any) => {
+        return rest.from(tableName).update(values)
+      },
+      delete: () => {
+        return rest.from(tableName).delete()
+      },
+      on: (event: string, callback: Function) => {
+        if (!this.realtime.isConnected()) {
+          this.realtime.connect()
+        }
+        return subscription.on(event, callback)
+      },
+    }
+    return builder
+  }
```

Comment:
I'm not entirely happy with this one. Is there a way to extend a function that is typescript friendly?

I tried this but Typescript didn't like it because it doesn't seem to pick up the that it is returning a `PostgrestClient` after I call `bind()`

```
  let rest = this._initPostgRESTClient()
  const builder = {
      rest,
      subscription,
      select: rest.from(tableName).select.bind(rest),
      insert: rest.from(tableName).insert.bind(rest),
      update: rest.from(tableName).update.bind(rest),
      delete: rest.from(tableName).delete.bind(rest)
    }
```

I think I need to do some with [TypeScript settings](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-2.html#strictbindcallapply)

Reply:
Doing the above gives me this error (also the case with the TS setting above turned on):

```
Return type of public method from exported class has or is using name 'PostgrestFilterBuilder' from external module [...] but cannot be named.
```

This kind of works:

```js
const builder = rest.from(tableName)
Object.assign(builder, {
  subscription,
  on: (event: 'INSERT' | 'UPDATE' | 'DELETE' | '*', callback: Function) => {
    if (!this.realtime.isConnected()) {
      this.realtime.connect()
    }
    return subscription.on(event, callback)
  },
})
```

But then there's no autocomplete for `on()`. I can't think of a way to do this without creating a new type, e.g. using [mixin](https://www.typescriptlang.org/docs/handbook/mixins.html).

Reply:
OK cool - I'm going to try exporting the `PostgrestQueryBuilder` from `postgrest-js`, then I will extend it with the `on()` function.

I have a few other things to tidy up here which I'm working on now